### PR TITLE
Add StackedLinearFunctional and FlattenedLinearFunctional

### PR DESCRIPTION
**Pull request by timweiland**
The new inference engine relies on the FlattenedLinearFunctional, hence why I'm making the PR for these two now :)

### CODE REVIEW

File: src/linpde_gp/randprocs/crosscov/linfunctls/_stacked.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    def _evaluate_jax(self, x: jnp.ndarray) -> jnp.ndarray:
+        return jnp.concatenate((self._kL1a(x), self._kL2a(x)), axis=-1)
+
+@linfunctls.StackedLinearFunctional.__call__.register(  # pylint: disable=no-member
+    CovarianceFunction_Identity_Stacked
+)
+def _(self, kSa: CovarianceFunction_Identity_Stacked, /) -> pn.linops.LinearOperator:
+    L1kL1a = pn.linops.Matrix(self.L1(kSa.kL1a))
+    L1kL1a.is_symmetric = True
+    L1kL1a.is_positive_definite = True
```
Comment (marvinpfoertner) [inline src/linpde_gp/randprocs/crosscov/linfunctls/_stacked.py:77]: I don't think that this is true in general.
Isn't it possible that $k\mathcal{L}^*$ used a different linear functional, i.e. not `L1`?

File: tests/linpde_gp/linfunctls/test_flattened.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    L_res = L(f)
+
+    assert L_res.shape == D_res.shape
+    np.testing.assert_allclose(L_res, D_res)
+
+def test_tensor_becomes_flat():
+    I, O = 2, 3
+    N = 5
+    X = np.zeros((N, I))
+    f = linpde_gp.functions.Zero((I,), (O,))
```
Comment (marvinpfoertner) [inline tests/linpde_gp/linfunctls/test_flattened.py:24]: Can we use a nonzero `Constant` function here? Something like a reshaped `arange` would reveal more errors

**Review by marvinpfoertner — COMMENTED** [review]
Thanks for the PR! 🙏🏻

I loosely follow the convention of using argument and property names like `linfctl` for "inner"/"wrapped" linear functionals and operators. Can you apply this convention to the PR? As an example, I already made a couple of detailed suggestions for the `FlattenedLinearFunctional`.

File: src/linpde_gp/linfunctls/_flattened.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import functools
+
+import numpy as np
+import probnum as pn
+
+from ._linfunctl import LinearFunctional
+
+class FlattenedLinearFunctional(LinearFunctional):
```
[the remark this answers is not in the document]
Comment (timweiland) [reply src/linpde_gp/linfunctls/_flattened.py]: I still find `FlattenedLinearFunctional` slightly more intuitive because even with a different order, the goal is still to turn objects with multiple dimensions into flat objects with a single dimension. But I'm also fine with `ReshapedLinearFunctional` :)

File: src/linpde_gp/randprocs/crosscov/linfunctls/_flattened.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    if not pv_crosscov.reverse:
+        # k(x, .)
+        target_shape = self.output_shape + inner_shape[inner_output_ndim:]
+    else:
+        # k(., x)
+        target_shape = inner_shape[:-inner_output_ndim] + self.output_shape
+    return inner_result.reshape(target_shape, order="C")
+
+
+class CovarianceFunction_Identity_Flattened(ProcessVectorCrossCovariance):
```
[the remark this answers is not in the document]
Comment (timweiland) [reply src/linpde_gp/randprocs/crosscov/linfunctls/_flattened.py]: Great idea! This also ended up simplifying the code a lot.

File: src/linpde_gp/randprocs/crosscov/linfunctls/_dirac.py [hunk 5]
```diff
@@ -165,13 +165,9 @@ def __init__(
...
         self._dirac = dirac
 
-        randproc_output_shape = self._covfunc.output_shape[
-            self._dirac.input_codomain_ndim :
-        ]
-
         super().__init__(
             randproc_input_shape=self._covfunc.input_shape,
-            randproc_output_shape=randproc_output_shape,
+            randproc_output_shape=self._covfunc.output_shape_1,
```
[the remark this answers is not in the document]
Comment (timweiland) [reply src/linpde_gp/randprocs/crosscov/linfunctls/_dirac.py:170]: It does! You already changed it. I think you just forgot to change it for this one :)

File: src/linpde_gp/randprocs/crosscov/linfunctls/_stacked.py (hunk 1 above)
Comment (timweiland) [reply src/linpde_gp/randprocs/crosscov/linfunctls/_stacked.py:77]: You're right, I didn't consider that.

File: src/linpde_gp/linfunctls/_stacked.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    def L1(self) -> FlattenedLinearFunctional:
+        return self._L1
+    
+    @property
+    def L2(self) -> FlattenedLinearFunctional:
+        return self._L2
+
+    @functools.singledispatchmethod
+    def __call__(self, f, /, **kwargs):
+        return np.concatenate((self._L1(f), self._L2(f)))
```
[the remark this answers is not in the document]
Comment (timweiland) [reply src/linpde_gp/linfunctls/_stacked.py]: I'm not sure. Maybe we should discuss the purpose of `StackedLinearFunctional`?
My original idea was that it is simply a utility to combine different linear functionals - even if they potentially have different output shapes. So even linear functionals with shapes (20, 40) and (10, 50, 80) would be stackable. To make this work internally, I would wrap those linear functionals with a `ReshapedLinearFunctional` to make their outputs flat. In this case, I don't see the advantage of using `np.stack`, adding a new dummy dimension, over using `np.concatenate`, which keeps everything flat.

But I'm not sure if it's even that easy, especially once we start integrating KeOps. What do you think?

[the remark this answers is not in the document]
Comment (marvinpfoertner) [reply src/linpde_gp/linfunctls/_stacked.py]: Ah, I see. In this case I would suggest we use the name `ConcatenatedLinearFunctional` and stick to the `np.concatenate` internally, since there might be cases where we actually need a `StackedLinearFunctional`.

**Comment by timweiland** [comment]
Closing this for now since - due to the introduction of the `linop` method - we don't need it at the moment.
